lib/logproto/logproto-client.h

```diff
@@ -123,39 +124,53 @@ log_proto_client_handshake(LogProtoClient *s, gboolean *handshake_finished)
 }
 
 static inline gboolean
-log_proto_client_poll_prepare(LogProtoClient *s, gint *fd, GIOCondition *cond, gint *timeout)
+log_proto_client_poll_prepare(LogProtoClient *self, GIOCondition *cond, GIOCondition *idle_cond, gint *timeout)
 {
-  gboolean result = s->poll_prepare(s, fd, cond, timeout);
+  GIOCondition transport_cond = 0;
+  gboolean result = TRUE;
+
+  if (log_transport_stack_poll_prepare(&self->transport_stack, &transport_cond))
+    goto exit;
+
+  result = self->poll_prepare(self, cond, idle_cond, timeout);
+
+  /* Protos are not allowed to set both IN and OUT because of "transport_reversed_io_direction" and TLS requirements:
+   * "If you get SSL_ERROR_WANT_WRITE from SSL_write() then you should not do any other operation that could trigger
+   * IO other than to repeat the previous SSL_write() call."
+   */
+  g_assert(!((*cond & G_IO_IN) && (*cond & G_IO_OUT)));
```

Comment:
@bazsi This is the consequence of moving everything into the abstract class and trying to use the bool logic  you suggested.

I think my original logic was more allowing. I agree it was ugly, but it is not possible to be transport-transparent in the protos without those callbacks we talked about.